tools/ktx/command_encode_astc.cpp

```diff
@@ -0,0 +1,205 @@
+    </dl>
+    @snippet{doc} ktx/compress_utils.h command options_compress
+    @snippet{doc} ktx/command.h command options_generic
+
+@section ktx_encode_astc_exitstatus EXIT STATUS
+    @snippet{doc} ktx/command.h command exitstatus
+
+@section ktx_encode_astc_history HISTORY
+
+@par Version 4.0
+ - Initial version
+
+@section ktx_encode_astc_author AUTHOR
+    - Wasim Abbas, Arm Ltd. www.arm.com
+*/
+class CommandEncodeAstc : public Command {
+    enum {
+        all = -1,
+    };
+
+    struct OptionsEncode {
+        void init(cxxopts::Options& opts);
+        void process(cxxopts::Options& opts, cxxopts::ParseResult& args, Reporter& report);
+    };
+
+    Combine<OptionsEncode, OptionsASTC, OptionsMetrics, OptionsCompress, OptionsSingleInSingleOut, OptionsGeneric> options;
+
+public:
+    virtual int main(int argc, char* argv[]) override;
+    virtual void initOptions(cxxopts::Options& opts) override;
+    virtual void processOptions(cxxopts::Options& opts, cxxopts::ParseResult& args) override;
+
+private:
+    void executeEncodeAstc();
+};
+
+// -------------------------------------------------------------------------------------------------
+
+int CommandEncodeAstc::main(int argc, char* argv[]) {
+    try {
+        parseCommandLine("ktx encode-astc",
+                "ASTC Encode the KTX file specified as the input-file argument,\n"
+                "    optionally supercompress the result, and save it as the output-file.",
+                argc, argv);
+        executeEncodeAstc();
+        return +rc::SUCCESS;
+    } catch (const FatalError& error) {
+        return +error.returnCode;
+    } catch (const std::exception& e) {
+        fmt::print(std::cerr, "{} fatal: {}\n", commandName, e.what());
+        return +rc::RUNTIME_ERROR;
+    }
+}
+
+void CommandEncodeAstc::OptionsEncode::init(cxxopts::Options&) {
+}
+
+void CommandEncodeAstc::OptionsEncode::process(cxxopts::Options&, cxxopts::ParseResult&, Reporter&) {
+}
+
+void CommandEncodeAstc::initOptions(cxxopts::Options& opts) {
+    options.init(opts);
+}
+
+void CommandEncodeAstc::processOptions(cxxopts::Options& opts, cxxopts::ParseResult& args) {
+    options.process(opts, args, *this);
+}
+
+void CommandEncodeAstc::executeEncodeAstc() {
+    InputStream inputStream(options.inputFilepath, *this);
+    validateToolInput(inputStream, fmtInFile(options.inputFilepath), *this);
+
+    KTXTexture2 texture{nullptr};
+    StreambufStream<std::streambuf*> ktx2Stream{inputStream->rdbuf(), std::ios::in | std::ios::binary};
+    auto ret = ktxTexture2_CreateFromStream(ktx2Stream.stream(), KTX_TEXTURE_CREATE_LOAD_IMAGE_DATA_BIT, texture.pHandle());
+    if (ret != KTX_SUCCESS)
+        fatal(rc::INVALID_FILE, "Failed to create KTX2 texture: {}", ktxErrorString(ret));
+
+    if (texture->supercompressionScheme != KTX_SS_NONE)
+        fatal(rc::INVALID_FILE, "Cannot encode KTX2 file with {} supercompression.",
+            toString(ktxSupercmpScheme(texture->supercompressionScheme)));
```

Comment:
This is too blunt a hammer. Only BasisLZ supercompressed images should be rejected. There is no reason to reject uncompressed images supercompressed with zstd or zlib.

And as is being done by the switch below this comment UASTC compressed images and any `*_BLOCK` VkFormats should be rejected as any such encode-decode sequence will add noise plus we don't have decoders for many of the formats.

Reply:
I must say I haven't started looking into these options in detail yet. As you might know this is a straight copy from `command_encode.cpp`. Does this option not make sense there https://github.com/KhronosGroup/KTX-Software/blob/97feeae15db211fab5798174f4d2d03b08e428ef/tools/ktx/command_encode.cpp#L156 either?

Also glad your reviewed this, does it mean the other checks in `executeEncodeAstc` makes sense?

Reply:
Assuming this also requires an `inflat` from whatever supercompression is provided before it can be sent to the encoder? At least astc encoder doesn't support suprecompressed input.

Reply:
Inflate is done automatically by libktx when a ktxTexture2 is created from a file.

command_encode.cpp should be refusing to anything with a file already supercompressed with BasisLZ. For other supercompression schemes it should open the file to check the format and refuse if it is UASTC, any block-compressed GPU format or if it is not one of the small number of suitable input formats. The last 2 checks apply too when the input is not supercompressed.

Reply:
>Inflate is done automatically by libktx when a ktxTexture2 is created from a file.

Let me get that right. If I load a ktx2 file in `encode-astc` that was supercompressed. It will be inflated but `texture->supercompressionScheme` will still tell me what supercompression it was?

I think I am getting confused again with the "supercompression" types we are referring to in here.
The only things that talk about "Supercompress" in the help doc for `create` and `encode` is in relation to `zstd` and `zlib` so I don't know what does the following mean?
> a file already supercompressed with BasisLZ

and

>For other supercompression schemes it should open the file to check the format and refuse if it is UASTC